docs/adr/009_lns_affinity.md

```diff
@@ -0,0 +1,244 @@
+
+- Deduplication strategies Mark and None: these strategies rely on multiple LNSs sending message.
+Potentially we could consider other work arounds for the IoT Hub limitation of a single connection
+per device but we find it acceptable for the Mark and None strategies to not be as scalable as the
+Drop strategy and will only document this limitation for potential users to be aware of.
+
+- LNS performs operations on behalf of a device/sensor and a concentrator/station. However since a
+concentrator can be connected to at most one LNS, there is no ping-pong happening with operations on
+stations.
+
+## Limitations
+
+- We should avoid introducing additional calls to the Function as this would also hurt scalability.
+  
+## In-scope
+
+- The problem can be manifested whenever we do operations against Iot Hub on behalf of edge devices.
+  These can be:
+  - Twin reads
+  - Twin writes (updates/deletes)
+  - D2C messages
+  - C2D messages
+- Roaming edge devices (that potentially lose connectivity to an LNS) are in scope.
+
+### Problematic connections opened on behalf of edge devices
+
+- Data message flows
+  - Join: 
+  - Data: MessageDispatcher.DispatchLoraDataMessage gets the twin (if LoRaDevice not in cache)
+    - Direct method for Class C should return an error if it could not send
+      message downlink so that the user can retry.
+
+Version, LNS discovery and CUPS update endpoints are not affected.
+
+### Background tasks
+
+- LoRaDeviceCache refresh in the background: every 10 minutes we are checking for devices that need
+  refreshing (those that were not refreshed in the last 2 days). Refreshing a device gets the twin
+  of the device ❔
+  - if we are the losing LNS we adjust LastSean property but not refresh the entry
+  - in order to not have stale entries we could refresh the twin once we become the winning LNS from
+    the data (also required for join?) flow  
+
+Alternative:
+SearchByDevAddr: the function returns info about the winning gateway
+Requires doing the Mic computation on the Function
+
+## Possible solutions
+
+### Delayed processing of messages from losing LNSs
+
+The main idea here is to delay the processing of future messages for all gateways *besides* the
+winning one. This should give enough time to the chosen LNS to process the message and keep the
+active connection to Iot Hub. The winning gateway is decided by the Function.
+
+#### Example scenario in main data message flow
+
+Assuming the topology:
+
+```mermaid
+flowchart LR;
+    Device-->LBS1-->LNS1-->Function;
+    Device-->LBS2-->LNS2-->Function;
+    LNS1-->IoTHub;
+    LNS2-->IoTHub;
+```
+
+where Device sends data message A and then B.
+
+Here is a rundown of what should happen marked in **bold**:
+
+- Device sends first data message A.
+- We assume that LNS1 gets the message first. **LNS1 checks against
+  an in-memory dictionary DevEui -> flag** and since it has not seen this DevEui before (flag is
+  false) contacts the FunctionBundler.
```

Comment:
nit: FunctionBundler seems too much info here.

Reply:
and I think, if it is not in the cache, it will first talk to the GetDevice endpoint?

Reply:
I think I got it now. You would keep the state in that dictionary? But you would always call the function anyway no? There should be no additional action before the call?

Reply:
I would focus on the function side processing, rather than the LNS side delay.

Reply:
> I would focus on the function side processing, rather than the LNS side delay.

As discussed we think it's sufficient to use local delay on the LNS. Resolving this.